packages/aws-cdk/lib/api/cloudformation-deployments.ts

```diff
@@ -372,6 +371,101 @@ export class CloudFormationDeployments {
     return stack.exists;
   }
 
+  private async prepareSdkWithLookupOrDeployRole(stackArtifact: cxapi.CloudFormationStackArtifact): Promise<ISDK> {
+    // try to assume the lookup role
+    try {
+      const result = await prepareSdkWithLookupRoleFor(this.sdkProvider, stackArtifact);
+      if (result.didAssumeRole) {
+        return result.sdk;
+      }
+    } catch { }
+    // fall back to the deploy role
+    return (await this.prepareSdkFor(stackArtifact, undefined, Mode.ForReading)).stackSdk;
+  }
+
+  private async readCurrentStackTemplate(stackName: string, stackSdk: ISDK) : Promise<Template> {
+    const cfn = stackSdk.cloudFormation();
+    const stack = await CloudFormationStack.lookup(cfn, stackName);
+    return stack.template();
+  }
+
+  private async addNestedTemplatesToGeneratedAndDeployedStacks(
+    rootStackArtifact: cxapi.CloudFormationStackArtifact,
+    generatedParentTemplate: any,
+    deployedParentTemplate: any,
+    parentStackName: string | undefined,
+    sdk: ISDK,
+  ): Promise<void> {
+    const listStackResources = parentStackName ? new LazyListStackResources(sdk, parentStackName) : undefined;
+    for (const [nestedStackLogicalId, generatedNestedStackResource] of Object.entries(generatedParentTemplate.Resources ?? {})) {
+      if (!this.isCdkManagedNestedStack(generatedNestedStackResource)) {
+        continue;
+      }
+
+      const assetPath = generatedNestedStackResource.Metadata['aws:asset:path'];
+      const nestedStackTemplates = await this.getNestedStackTemplates(
+        rootStackArtifact, assetPath, nestedStackLogicalId, parentStackName, listStackResources, sdk,
+      );
+
+      generatedNestedStackResource.Properties.NestedTemplate = nestedStackTemplates.generatedNestedTemplate;
+
+      deployedParentTemplate.Resources = deployedParentTemplate.Resources ?? {};
+      const deployedNestedStackResource = deployedParentTemplate.Resources[nestedStackLogicalId] ?? {};
+      deployedParentTemplate.Resources[nestedStackLogicalId] = deployedNestedStackResource;
+      deployedNestedStackResource.Type = 'AWS::CloudFormation::Stack';
+      deployedNestedStackResource.Properties = deployedNestedStackResource.Properties ?? {};
+      deployedNestedStackResource.Properties.NestedTemplate = nestedStackTemplates.deployedNestedTemplate;
+
+      await this.addNestedTemplatesToGeneratedAndDeployedStacks(
+        rootStackArtifact,
+        nestedStackTemplates.generatedNestedTemplate,
+        nestedStackTemplates.deployedNestedTemplate,
+        nestedStackTemplates.nestedStackName,
```

Comment:
How about we make `addNestedTemplatesToGeneratedAndDeployedStacks()` just take a `NestedStackTemplates` as an argument, instead of its three properties separately? I think that will simplify the code also.

Reply:
I like this idea. Should we change `NestedStackTemplates` to be `TemplatePair`, or something similar (to indicate that it contains both the generated and deployed templates) and to remove the `nested` from all of its member names?

Reply:
Sure! Something like:

```ts
interface StackTemplates {
  readonly generatedTemplate: any;
  readonly deployedTemplate: any;
  readonly stackName?: string;
}
```

should work!

Reply:
Maybe even `deployedStackName` instead of `stackName` (I _think_ that's what this property actually represents).

Reply:
My bad, I had deleted that comment because I wasn't sure of the suggestion.